File: Week_03/subsets/Solution.java
package subsets;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * 回溯
 * Time：O(2^n)，n个格子，每个格子有 2 种选择
 * Space：O(n)，递归的最深度，O(2^n)结果的个数
 */
class Solution2 {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        dfs(nums, ans, 0, nums.length, new ArrayList<>());
        return ans;
    }
    private void dfs(int[] nums, List<List<Integer>> ans, int i, int n, List<Integer> list) {
        if (i == n) {
            ans.add(new ArrayList<>(list));
            return;
        }
        //选择放
        list.add(nums[i]);
        dfs(nums, ans, i + 1, n, list);
        //选择不放
        list.remove(list.size() - 1);
        dfs(nums, ans, i + 1, n, list);
    }
}

/**
 * 循环
 */
class Solution3 {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        res.add(Collections.EMPTY_LIST);
        for (int num : nums) {
            List<List<Integer>> temp = new ArrayList<>();
            for (List<Integer> list : res) {
                ArrayList<Integer> newList = new ArrayList<>(list);
                newList.add(num);
                temp.add(newList);
            }
            res.addAll(temp);
        }
        return res;
    }
    public static void main(String[] args) {
        List<List<Integer>> subsets = new Solution3().subsets(new int[]{1, 2, 3});
        System.out.println(subsets);
    }
}

File: Week_03/subsets/Solution.py
from typing import List

# 递归回溯
# Time: O(2^n), 每个元素有两种选择，放或者不放
# Space: O(2^n)
class Solution1:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        res = []

        def backtracking(i=0, path=[]):
            if i == n:
                res.append(path[:]) # 拷贝保存
                return
            path.append(nums[i])
            backtracking(i + 1, path)  # 放元素
            path.pop()
            backtracking(i + 1, path)  # 不放元素

        backtracking()
        return res

print(Solution1().subsets([1,2,3]))

# 循环，从第一个空子集开始，每次向子集添加新的整数
class Solution2:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = [[]]
        for num in nums:
            res += (cur + [num] for cur in res)
        return res
